File: hadesc/src/main/kotlin/hadesc/ast/expression.kt
package hadesc.ast

import hadesc.location.HasLocation
import hadesc.location.SourceLocation

sealed class Expression : HasLocation {
    data class Error(override val location: SourceLocation) : Expression()
    data class Var(
        val name: Identifier
    ) : Expression() {
        override val location: SourceLocation
            get() = name.location
    }

    data class Call(
        override val location: SourceLocation,
        val callee: Expression,
        val args: List<Arg>
    ) : Expression()

    data class Property(
        override val location: SourceLocation,
        val lhs: Expression,
        val property: Identifier
    ) : Expression()

    data class ByteString(
        override val location: SourceLocation,
        val bytes: ByteArray
    ) : Expression()

    data class BoolLiteral(
        override val location: SourceLocation,
        val value: Boolean
    ) : Expression()
}

data class Arg(
    val expression: Expression
) : HasLocation {
    override val location: SourceLocation
        get() = expression.location
}

File: lib/ast/Expression.cpp
//
// Created by dhruv on 07/08/20.
//
#include "hades/ast/Expression.h"

namespace hades {

auto Expression::location() const -> const SourceLocation & {
  return m_location;
}

Expression::Expression(SourceLocation location, Kind kind) noexcept
    : m_location{location}, m_kind{kind} {}

IntLiteral::IntLiteral(SourceLocation location, i64 value) noexcept
    : Expression(location, Kind::INT_LITERAL), m_value{value} {}

VarExpression::VarExpression(Identifier name) noexcept
    : Expression(name.location(), Kind::VAR), m_name(name) {}

Arg::Arg(Optional<Identifier> label, const Expression *value) noexcept
    : m_label(std::move(label)), m_value(value) {}

auto Arg::value() const -> const Expression & { return *m_value; }

auto Arg::label() const -> const Optional<Identifier> & { return m_label; }

auto Arg::location() const -> SourceLocation {
  if (label().hasValue()) {
    return SourceLocation::between(&label().getValue(), &value());
  } else {
    return value().location();
  }
}

Call::Call(SourceLocation location, const Expression *callee,
           Call::Args &&args) noexcept
    : Expression(location, Kind::CALL), m_callee(callee),
      m_args(std::move(args)) {}

auto Call::args() const -> const Args & { return m_args; }

auto Call::callee() const -> const Expression & { return *m_callee; }

} // namespace hades